Add lint `clones_into_boxed_slices`

aedank0 opened the pull request:
Adds new lint `clones_into_boxed_slices` which detects clone-like functions followed by `into_boxed_...` instead of using `Box::from(...)`.

Closes rust-lang/rust-clippy#15951, closes rust-lang/rust-clippy#15373

changelog: new lint: [`clones_into_boxed_slices`]

samueltardieu commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+    let mut sugg = Sugg::hir_with_context(cx, inner, full_span.ctxt(), placeholder, &mut applicability);
+
+    let inner_ty = cx.typeck_results().expr_ty(inner);
+    let mut ref_count = count_refs(inner_ty);
```
You can use `peel_and_count_ty_refs` from `clippy_utils::ty`:

```suggestion
    let mut ref_count = peel_and_count_ty_refs(inner_ty).1 as isize;
```

samueltardieu commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+        || ty.is_str()
+        || ty.is_diag_item(cx, sym::cstr_type)
+        || ty.is_diag_item(cx, sym::Path)
+        || ty.is_diag_item(cx, sym::OsStr)
```
You should avoid doing multiple lookups:


```suggestion
    ty.is_slice() || ty.is_str() || matches!(ty.opt_diag_name(cx.tcx), Some(sym::cstr_type | sym::Path | sym::OsStr))
```

samueltardieu commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+                sym::into_boxed_c_str,
+                sym::into_boxed_os_str,
+            ]
+            .contains(&second_method_path.ident.name)
```
Matching by name is not enough and can lead to false positives. You should ensure that this is an inherent method call and that the receiver type is the one you expect.

As an alternative, since those methods are not diag items, you could add them to `clippy_utils::paths`.

samueltardieu commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+                    if second_method_path.ident.name == sym::into_boxed_path && !inner_ty.is_diag_item(cx, sym::Path) {
+                        // PathBuf's from(...) can convert from other str types,
+                        // so Path::new(...) must be used to assure resulting Box is the correct type
+                        show_lint(cx, full_span, arg, true, Some("Path::new("), "...", Some(")"));
```
`show_lint()` should know that this suggestion should be degraded to `MaybeIncorrect` at most, because `Path` may not be have been imported.

samueltardieu reviewed (changes requested):
Interesting lint. It should probably be located inside the `methods` directory though.

Also, it should check the MSRV to ensure that the proposed fixes are actionable. For example `.as_c_str()` didn't exist before Rust 1.20.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/16095/62589a21d351f63d77dcd173c31107e9f1d06328..1ef6ca975017fe207a89fb4f6df79794e479f2b9)*

llogiq commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+) {
+    let mut applicability = Applicability::MachineApplicable;
+
+    while let ExprKind::AddrOf(BorrowKind::Ref, _, expr) | ExprKind::Unary(UnOp::Deref, expr) = inner.kind {
```
What if someone puts curly brackets around a subexpression, such as `&{ foo.to_string() }`? That would be an `ExprKind::Block` with one return expression, defeating this loop.

llogiq commented on clippy_lints/src/clones_into_boxed_slices.rs:
```diff
@@ -0,0 +1,204 @@
+impl<'tcx> LateLintPass<'tcx> for ClonesIntoBoxedSlices {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, second_method: &'tcx Expr<'_>) {
+        // Is the second method into_boxed_...?
+        if let ExprKind::MethodCall(second_method_path, first_method, _, _) = second_method.kind
```
This would certainly be simplified by putting this lint into `methods`. The matching is basically already done there.

aedank0 commented:
@rustbot ready

Thanks for the feedback! I moved the lint into `methods` and fixed the other issues.
I messed up the rebase so I had to fix it and force push. First time contributing to an open-source project, my apologies.

For the MSRV, almost all of the `into_boxed...` methods are added on or after the version that the methods used in the suggestions were. The only exception seems to be `into_boxed_str` and `as_str`, which were added in 1.4.0 and 1.7.0 respectively, so I set the MSRV to be 1.7.0.

Also, I wasn't sure how to check if the `into_boxed_...` method calls were inherent or not, but it does now check for their respective receiver types.
